presto-hive/src/main/java/com/facebook/presto/hive/orc/OrcSelectivePageSourceFactory.java

```diff
@@ -0,0 +1,727 @@
+                .ifPresent(filterFunctions::add);
+
+        if (TRUE_CONSTANT.equals(filter)) {
+            return filterFunctions.build();
+        }
+
+        DynamicFilterExtractResult extractDynamicFilterResult = extractDynamicFilters(filter);
+
+        // dynamic filter will be added through subfield pushdown
+        filter = and(extractDynamicFilterResult.getStaticConjuncts());
+
+        if (!isAdaptiveFilterReorderingEnabled(session)) {
+            filterFunctions.add(new FilterFunction(session.getSqlFunctionProperties(), determinismEvaluator.isDeterministic(filter), predicateCompiler.compilePredicate(session.getSqlFunctionProperties(), session.getSessionFunctions(), filter).get()));
+            return filterFunctions.build();
+        }
+
+        List<RowExpression> conjuncts = extractConjuncts(filter);
+        if (conjuncts.size() == 1) {
+            filterFunctions.add(new FilterFunction(session.getSqlFunctionProperties(), determinismEvaluator.isDeterministic(filter), predicateCompiler.compilePredicate(session.getSqlFunctionProperties(), session.getSessionFunctions(), filter).get()));
+            return filterFunctions.build();
+        }
+
+        // Use LinkedHashMap to preserve user-specified order of conjuncts. This will be the initial order in which filters are applied.
+        Map<Set<Integer>, List<RowExpression>> inputsToConjuncts = new LinkedHashMap<>();
+        for (RowExpression conjunct : conjuncts) {
+            inputsToConjuncts.computeIfAbsent(extractInputs(conjunct), k -> new ArrayList<>()).add(conjunct);
+        }
+
+        inputsToConjuncts.values().stream()
+                .map(expressions -> binaryExpression(AND, expressions))
+                .map(predicate -> new FilterFunction(session.getSqlFunctionProperties(), determinismEvaluator.isDeterministic(predicate), predicateCompiler.compilePredicate(session.getSqlFunctionProperties(), session.getSessionFunctions(), predicate).get()))
+                .forEach(filterFunctions::add);
+
+        return filterFunctions.build();
+    }
+
+    private static Set<Integer> extractInputs(RowExpression expression)
+    {
+        ImmutableSet.Builder<Integer> inputs = ImmutableSet.builder();
+        expression.accept(new InputReferenceBuilderVisitor(), inputs);
+        return inputs.build();
+    }
+
+    private static class InputReferenceBuilderVisitor
+            extends DefaultRowExpressionTraversalVisitor<ImmutableSet.Builder<Integer>>
+    {
+        @Override
+        public Void visitInputReference(InputReferenceExpression input, ImmutableSet.Builder<Integer> builder)
+        {
+            builder.add(input.getField());
+            return null;
+        }
+    }
+
+    private static String splitError(Throwable t, Path path, long start, long length)
+    {
+        return format("Error opening Hive split %s (offset=%s, length=%s): %s", path, start, length, t.getMessage());
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

This code uses &#39;%s&#39; to format long: length (declared at line 683), start (declared at line 683). This is a potential locale-sensitive handling issue. It might cause errors in the handling and processing of the statement at line: 685. Consider formatting this data with &#39;%d&#39; instead.